### PR TITLE
Fix an issue with searching insertion index in NSArray.

### PR DESCRIPTION
**Pull request by bSr43**
This PR implements the behavior that we find on OS X when searching for an insertion index in an NSArray.

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
@ms-jihua is added to the review. #Closed

File: Frameworks/Foundation/NSArray.mm [hunk 1]
```diff
@@ -665,6 +665,10 @@ - (NSUInteger)indexOfObject:(NSObject*)obj
                                             _CFComparatorFunctionFromComparator,
                                             &comparator);
 
+	if (index >= [self count] && (options & NSBinarySearchingInsertionIndex)) {
+		return index;
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSArray.mm]: nit: please run clang-format with the .clang-format file in the root of our repo; we're a 4-space-indent codebase! :)

Comment (bSr43) [inline Frameworks/Foundation/NSArray.mm]: Oups... sorry about that :)

File: tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm [hunk 2]
```diff
@@ -172,6 +172,12 @@ static int objectIndexInArray(NSArray* array, int value, int startingFrom, int l
         @"If NSBinarySearchingInsertionIndex is specified and no other options provided NSArray returns any equal or one larger "
         @"index than any matching object’s index.");
 
+    auto anyIndexToInsertNineteen = objectIndexInArray(array, 19, 0, array.count, NSBinarySearchingInsertionIndex);
+    ASSERT_TRUE_MSG(
+        anyIndexToInsertNineteen == array.count,
+        @"If NSBinarySearchingInsertionIndex is specified and the value is greater than any objects of the array, the returned insertion index is at the end of the array"
+        @"index than any matching object’s index.");
```
Comment (bbowman) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm]: The second @"" is a continuation of the sentence and, to me, does not lexically parse well. It at least needs correct spacing... Was this just copied from above?

**Comment by bbowman** [comment]
:clock1:

File: tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm (hunk 2 above)
Comment (bSr43) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm]: You're right, just a stupid copy & paste... :/

**Comment by msft-Jeyaram** [comment]
:shipit:

File: tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm [hunk 3]
```diff
@@ -172,6 +172,11 @@ static int objectIndexInArray(NSArray* array, int value, int startingFrom, int l
         @"If NSBinarySearchingInsertionIndex is specified and no other options provided NSArray returns any equal or one larger "
         @"index than any matching object’s index.");
 
+    auto anyIndexToInsertNineteen = objectIndexInArray(array, 19, 0, array.count, NSBinarySearchingInsertionIndex);
+    ASSERT_TRUE_MSG(
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm]: It looks like this ended up in the ported Swift tests (ReferenceFoundation directory); would you mind moving it to TestNSArray.mm in the parent directory?
We're trying to keep the tests from the reference platform separate from the ones we've written. :smile:
Thanks!

Comment (DHowett-MSFT) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm]: Sorry, that's `NSArrayTests.mm`, not `TestNSArray.mm`. Huh.

---
In reply to: [67031200](https://github.com/Microsoft/WinObjC/pull/557#discussion_r67031200) [](ancestors = 67031200)

**Comment by jaredhms** [comment]
:shipit:

File: tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm (hunk 3 above)
Comment (bSr43) [inline tests/unittests/Foundation/ReferenceFoundation/TestNSArray.mm]: Sure! I moved the tests regarding the NSBinarySearchingInsertionIndex option in the NSArrayTests.mm file.

**Comment by DHowett-MSFT** [comment]
:shipit:

**Comment by DHowett-MSFT** [comment]
Thanks for your contribution, @bSr43!

**Pull request merged**